### ISSUE TITLE
Is it possible to set font while filling fields of for using PDFWriter class?

### ISSUE THREAD

**Issue by paritoshranjan** [py-pdf/pypdf#2253]
I am filling a form using PDFWriter class. The form is getting field.

Can the font type and font size of the values be specified while filling the form? If yes, how?

**Comment by pubpub-zz** [comment]
Work is in progress for this new feature

**Comment by stefan6419846** [comment]
@PJBrs Are you able to extend our docs to include an example which covers this use case?

**Comment by PJBrs** [comment]
I guess I can...

Just one question, perhaps it might help to look at the current method signature:
```
    def update_page_form_field_values(
        self,
        page: Union[PageObject, list[PageObject], None],
        fields: Mapping[str, Union[str, list[str], tuple[str, str, float]]],
        flags: FA.FfBits = FFBITS_NUL,
        auto_regenerate: Optional[bool] = True,
        flatten: bool = False,
    ) -> None:
        """
        Update the form field values for a given page from a fields dictionary.

        Copy field texts and values from fields to page.
        If the field links to a parent object, add the information to the parent.

        Args:
            page: `PageObject` - references **PDF writer's page** where the
                annotations and field data will be updated.
                `List[Pageobject]` - provides list of pages to be processed.
                `None` - all pages.
            fields: a Python dictionary of:

                * field names (/T) as keys and text values (/V) as value
                * field names (/T) as keys and list of text values (/V) for multiple choice list
                * field names (/T) as keys and tuple of:
                    * text values (/V)
                    * font id (e.g. /F1, the font id must exist)
                    * font size (0 for autosize)

            flags: A set of flags from :class:`~pypdf.constants.FieldDictionaryAttributes.FfBits`.

            auto_regenerate: Set/unset the need_appearances flag;
                the flag is unchanged if auto_regenerate is None.

            flatten: Whether or not to flatten the annotation. If True, this adds the annotation's
                appearance stream to the page contents. Note that this option does not remove the
                annotation itself.

        """
```

Font resource name and font size can now be passed as part of a tuple. Note in particular: `fields: Mapping[str, Union[str, list[str], tuple[str, str, float]]],`. This signature, to me, seems somewhat unfortunate for various reasons:

1. It doesn't appear that one can set font name / font size for multiple choice lists (might nog be applicable though, I don't know).
2. One can _either_ pass a string value _or_ a tuple of string, string, float, meaning that one _has_ to set both font name and font size.
3. Given the tuple signature for value, font name and font size, it is actually not so clear what one is passing (although something like the following is pretty self evident:
```
                writer.update_page_form_field_values(
                    writer.pages[0], {key: (fields[key], "/Kalam", 0)},
```

I don't know if this already is seen as accepted API that needs to be deprecated before it can be extended, but I guess something like the following would be much nicer:

        fields: Mapping[str, Union[str, list[str]], formatting_args**],

Where formatting_args = dict[font_name: Optional = None, font_size: Optional = None, font_color: Optional = None]

In sum, I can definitely add example code (the above is something I use myself. But if we allow ourselves a moment to reconsider the signature, we could also make the interface easier to use.

**EDIT** I know that my above suggestion doesn't work, but you'll probably understand what I mean :-)

**Comment by stefan6419846** [comment]
Having an explicit example in the docs is still nice/better, even if it could be derived from the docstring. Fro my experience, an explicit tutorial is something which is much more read and easier to understand than a direct docstring.

**Comment by PJBrs** [comment]
> Having an explicit example in the docs is still nice/better, even if it could be derived from the docstring. Fro my experience, an explicit tutorial is something which is much more read and easier to understand than a direct docstring.

I agree! I just wonder if the _current_ API is sufficiently clean to publicise it in our documentation, when, at least in my view, it can be made a lot easier...

**Comment by stefan6419846** [comment]
The API is already public and every breaking change would require a deprecation process anyway (with a corresponding example update). In which order documenting and improving makes sense remains to be discussed.

**Comment by PJBrs** [comment]
One way would to _add_ an API option for style option args while keeping the old API in place (albeit deprecated), Then we could add the new API to the documentation. But that would require an entirely new PR.

@stefan6419846 Please let me know which you prefer. I _can_ add some Documentation now and close this bug, or we keep it open, add a style api, document that, and then close this bug.

**Comment by stefan6419846** [comment]
This issue is old and uncritical enough to allow for delaying the documentation here until a "more intuitive" interface is available.

**Comment by PJBrs** [comment]
I was thinking along the lines of the following:

```
def update_page_form_field_values(
    self,
    page: Union[PageObject, list[PageObject], None],
    fields: Mapping[
        str, 
        Union[
            str,                                          # New/Old: Plain text
            list[str],                                    # New/Old: Plain choice list
            tuple[str, str, float],                       # OLD API: (value, font_name, font_size)
            tuple[Union[str, list[str]], dict[str, Any]]  # NEW API: (value, arbitrary_dict)
        ]
    ],
    flags: FA.FfBits = FFBITS_NUL,
    auto_regenerate: Optional[bool] = True,
    flatten: bool = False,
) -> None:
```

(with some help from Google Gemini).

In dict[str, Any], the use can add formatting suggestions that we parse in TextAppearanceStream.  In this way, we can immediately support users setting font name and font size, and adding further options later (such as font color, border size, border width, ...) would be very easy.

**Comment by stefan6419846** [comment]
While this approach looks like the most obvious consequence, I am not sure whether we are able to make this API more "structured" in a way that it is directly obvious what each value means and what values we can pass.

Specifically, I am thinking whether making the field value a container class with named attributes (for example a *dataclass*) would be more understandable than the current approach with arbitrary unnamed data types.

**Comment by PJBrs** [comment]
I had considered something like that, but it seemed to me that that would change the entire signature so much that all existing users need to move to using that new dataclass. I thought that the advantage of using a tuple of (value: Union[str, list[str]], FieldFormattingDataclass) is that existing use cases that do not add field formatting can keep using the old api of passing only the value Union[str, list[str]].  If you deprecate that as well, many more people will need to adapt.

**Comment by stefan6419846** [comment]
My main issue with your previous proposal most likely has been that it would introduce an arbitrary dictionary, which might not even be compatible with the actual supported parameters further down. Providing a container for these parameters (like `FieldFormattingDataclass` which you proposed) would solve this and indeed reduce the possible friction of reworking the whole API.
